Network/Socket/Types.hsc

```diff
@@ -83,6 +87,20 @@ instance Eq Socket where
     Socket ref1 _ == Socket ref2 _ = ref1 == ref2
 
 -- | Getting a file descriptor from a socket.
+--
+--   If a 'Socket' is shared with multiple threads and
+--   one uses 'fdSocket', unexpected issues may happen.
+--   Consider the following scenario:
+--
+--   1) Thread A acquires a 'Fd' from 'Socket' by 'fdSocket'.
+--
+--   2) Thread B close the 'Socket'.
+--
+--   3) Thread C opens a new 'Socket'. Unfortunately it gets the same 'Fd'
+--      number which thread A is holding.
+--
+--   In this case, it is safer for Thread A to clone 'Fd' by
+--   'System.Posix.IO.dup'.
```

Comment:
Note, that even `dup` is not safe if the socket is not locked against close between `fdSocket` and `dup`.  For a safe `dup` we'd need to provide an atomic variant, that is guaranteed to be duping the unclosed descriptor inside the given socket.  You'd need an MVar associated with the socket to implement a safe `dup`.

The solution is to avoid use of the same socket (other than acceptor or separation of read/write) in multiple threads. All other concurrency models are unsound, and no thread may manipulate a socket concurrently with another thread outside the limited exception cases.

Reply:
I will add this race condition to the doc.